### PR TITLE
Implement realMinMax() and minMax() methods to fix adapting interval performances 

### PR DESCRIPTION
**Pull request by tischi**
Work along the outline provided here: https://github.com/imglib/imglib2-roi/issues/60

### CODE REVIEW

**Comment by tischi** [comment]
@tpietzsch 
Could you have a look?
I think I did everything....

**Comment by tpietzsch** [comment]
I added a commit where I let `AbstractAdaptingInterval` extends `AbstractAdaptingRealInterval`, and implemented `realMinMax` in terms of `minMax`.
This will
* make the `instanceof` checks in `getMinMax` more robust, in case a `AbstractAdaptingInterval` is passed into the `getMinMax(RealInterval, double[], double[])` version.
* provide efficient implementations of the RealInterval methods.
 
I added stubs for all the `Interval` methods that should be overridden. Can you please add the implementations?

`RealTransformRealInterval` should be revised, relying more on the `realMinMax` method directly. As is is, that will be called many times., through the various realMin/Max methods. Could you have a look at that as well? I added a few TODO comments for getting started.

**Comment by tpietzsch** [comment]
Also `AbstractAdaptingInterval` should override `dimensions(Positionable)`.
I forgot that one...

**Comment by tischi** [comment]
Thanks @tpietzsch for having a look! I am truly impressed that you spotted

> in case a AbstractAdaptingInterval is passed into the getMinMax(RealInterval, double[], double[]) version

you have to explain me at some point how you find such stuff!

----

I can do all the requested changes, however not to screw anything up I would like to understand this part of the code:

https://github.com/imglib/imglib2-roi/blob/0f2177c818f5d48b934ef2ce8c569dcafef4833a/src/main/java/net/imglib2/roi/Bounds.java#L694

Frankly, I don't see how anything that happens in the while loop would affect the loop's conditional.
I mean, if `source` does not change (and I don't see how it would), the value of the variables that are testing in the while condition should not change?!

In other words, could this be rewritten without a while loop?  Probably I am overlooking something?!

**Comment by tpietzsch** [comment]
I think  the idea is that the `source` might be changed by another thread. But there is so many more problems if that actually happens... I think it is better to assume that this will be used single-threaded.

So yes, I think it should be rewritten without the while loop.

**Comment by tischi** [comment]
OK, I will rewrite it then. There are several things though that I don't get...

1. Isn't `transformToSource.numTargetDimensions() == source.numDimensions()`?
2. There are repeated calls to `this.transformToSource.numTargetDimensions()`. 
Can this change? If not, why not just have a field `numSourceDimensions = transformToSource.numTargetDimensions()` initialised in the constructor? I feel that  if the dimensions of `transformToSource.numTargetDimensions()` could change, then it maybe `transformToSource.numSourceDimensions()` could change and then also the initialization of the dimensions in the constructor `super(...)` would be wrong? 
3. If we would do `source.realMin( cachedSourceMin ); source.realMax( cachedSourceMax );` at the very beginning of `updateMinMax()` then we could reuse `cachedSourceMin` and `cachedSourceMax` later?!
4. Why is there no return statement after

**Comment by tischi** [comment]
(at least, i would only change `source.numDimensions()` once during the `updateMinMax()`...)

**Comment by tpietzsch** [comment]
First of all, this isn't originally my code, and it's been a few years since I reviewed it... So I'm also guessing a bit.

> 1. Isn't `transformToSource.numTargetDimensions() == source.numDimensions()`?

Yes, let's assume that!

(In principle, I think that you might want to be able to have a 3D source and apply a 2D transformation that leaves the 3rd dimension untouched. But that wouldn't work with the current code. And I think it would make things a lot more complicated for marginal benefit).

> 2. There are repeated calls to `this.transformToSource.numTargetDimensions()`.
>    Can this change? If not, why not just have a field `numSourceDimensions = transformToSource.numTargetDimensions()` initialised in the constructor? I feel that  if the dimensions of `transformToSource.numTargetDimensions()` could change, then it maybe `transformToSource.numSourceDimensions()` could change and then also the initialization of the dimensions in the constructor `super(...)` would be wrong?

Neither `transformToSource.numTargetDimensions()` nor `numSourceDimensions()` should ever change.
So yes, it's a good idea to make a field `numSourceDimensions = transformToSource.numTargetDimensions()` initialised in the constructor.

> 3. If we would do `source.realMin( cachedSourceMin ); source.realMax( cachedSourceMax );` at the very beginning of `updateMinMax()` then we could reuse `cachedSourceMin` and `cachedSourceMax` later?!

Yes, sounds good.

> 4. Why is there no return statement after

no return statement after what?


Thanks for digging into it!!! 👍

File: src/main/java/net/imglib2/roi/Bounds.java [hunk 1]
```diff
@@ -801,88 +813,74 @@ public void realMinMax( double[] realMin, double[] realMax )
...
 				}
-
-				source.realMax( cachedSourceMax );
-				source.realMin( cachedSourceMin );
+				min[ d ] = minCorner;
+				max[ d ] = maxCorner;
 			}
 		}
 
 		private double[][] createCorners()
```
Comment (tischi) [inline src/main/java/net/imglib2/roi/Bounds.java:859]: @tpietzsch 
I don't understand the code here. If the code is good, shouldn't we put it as a static method somewhere? Something like `RealInterval ... = TransformHelper.estimateBounds(   RealTransform ..., RealInterval ... );`

**Comment by tischi** [comment]
@tpietzsch 

OK, I think I did everything:

1. https://github.com/imglib/imglib2-roi/pull/63/commits/e4a9054b60995844e4514b47e9d3e6910d769ff0
2. https://github.com/imglib/imglib2-roi/pull/63/commits/8333f92b4469aba11c022f8e9e01e8101aa08f6b
3. https://github.com/imglib/imglib2-roi/pull/63/commits/7e9dda103d9361e58602a4489c5a4898d24dcad1

Tests are passing.

I have one question, which I added as a comment to the last commit.

File: src/main/java/net/imglib2/roi/Bounds.java (hunk 1 above)
Comment (bogovicj) [reply src/main/java/net/imglib2/roi/Bounds.java:859]: > shouldn't we put it as a static method somewhere

I'd like that, as well as an implementation that would work for non-linear transformations too
https://github.com/imglib/imglib2-realtransform/pull/37

Comment (tpietzsch) [reply src/main/java/net/imglib2/roi/Bounds.java:858]: I refactored this a bit, and it should be clearer now what happens.

This method computes the corners of the source interval (see javadoc for explanation how)
https://github.com/imglib/imglib2-roi/blob/46e88901c684215b7da82cdd3518b0c56ba96e67/src/main/java/net/imglib2/roi/Bounds.java#L873-L892

Then this method takes the corners and puts them through the transformation
https://github.com/imglib/imglib2-roi/blob/46e88901c684215b7da82cdd3518b0c56ba96e67/src/main/java/net/imglib2/roi/Bounds.java#L858-L865

Finally, this method puts computes the bounding box of the transformed corners
https://github.com/imglib/imglib2-roi/blob/46e88901c684215b7da82cdd3518b0c56ba96e67/src/main/java/net/imglib2/roi/Bounds.java#L841-L855
